lib/pos_mempool_persister.go

```diff
@@ -0,0 +1,205 @@
+package lib
+
+import (
+	"bytes"
+	"github.com/deso-protocol/core/storage"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"sync"
+	"time"
+)
+
+const (
+	DbMempoolContextId = "transactions"
+	eventQueueSize     = 10000
+)
+
+type MempoolEventType int
+
+const (
+	MempoolEventAdd MempoolEventType = iota
+	MempoolEventRemove
+	MempoolEventExit
+)
+
+type MempoolPersisterStatus int
+
+const (
+	MempoolPersisterStatusRunning MempoolPersisterStatus = iota
+	MempoolPersisterStatusNotRunning
+)
+
+// MempoolPersister is responsible for persisting transactions in the mempool to the database. Whenever a transaction is
+// added or removed from the mempool, the MempoolPersister is notified via the OnMempoolEvent callback. The Persister
+// will then add the event to a queue. Periodically, the transaction queue is flushed to the database and all the cached
+// transactions are persisted.
+type MempoolPersister struct {
+	sync.Mutex
+	stopGroup  sync.WaitGroup
+	startGroup sync.WaitGroup
+
+	frequencyMilliseconds int
+	status                MempoolPersisterStatus
+	dbCtx                 *storage.DatabaseContext
+	eventQueue            chan *MempoolEvent
+	updateBatch           []*MempoolEvent
+}
+
+func NewMempoolPersister(dbCtx *storage.DatabaseContext, frequencyMilliseconds int) *MempoolPersister {
+	return &MempoolPersister{
+		frequencyMilliseconds: frequencyMilliseconds,
+		status:                MempoolPersisterStatusNotRunning,
+		dbCtx:                 dbCtx,
+		eventQueue:            make(chan *MempoolEvent, eventQueueSize),
+	}
+}
+
+func (mp *MempoolPersister) Start() {
+	if mp.status == MempoolPersisterStatusRunning {
+		return
+	}
+
+	mp.reset()
+	mp.stopGroup.Add(1)
+	mp.startGroup.Add(1)
+	go mp.run()
+	mp.startGroup.Wait()
+	mp.status = MempoolPersisterStatusRunning
+}
+
+func (mp *MempoolPersister) run() {
+	mp.startGroup.Done()
```

Comment:
Seems like you can just set the status to running here. If you do that, then the `startGroup` wait group becomes redundant and you can remove it. And all status changes happen within this function so they're guaranteed to be on the same thread
```suggestion
	mp.status = MempoolPersisterStatusRunning
```

Reply:
Ah this is an interesting little synchronization problem. Since `run()` is executed in a goroutine, it's possible that it'll not execute right away. If somehow we publish events to the persister before the `run()` goroutine gets scheduled, then these events will be lost. Having the startGroup ensures that the run loop is up and running.

Reply:
Ah, makes sense. It's worth leaving a comment here to explain that edge for future readers. Tricky stuff.